src/lib/vasi-sync/src/lazy_lock.rs

```diff
@@ -0,0 +1,176 @@
+use core::marker::PhantomData;
+use core::mem::MaybeUninit;
+use core::sync::atomic::{self, AtomicI8};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use vasi::VirtualAddressSpaceIndependent;
+
+use crate::sync::{ConstPtr, UnsafeCell};
+
+// For use with `LazyLock`. This trait is implemented for
+// `FnOnce` closures, so most users can just use that directly.
+// However in case they need to name a type in the LazyLock,
+// they can create a type that implements this trait instead.
+pub trait Producer<T> {
+    fn initialize(self) -> T;
+}
+
+impl<F, T> Producer<T> for F
+where
+    F: FnOnce() -> T,
+{
+    fn initialize(self) -> T {
+        (self)()
+    }
+}
+
+#[derive(TryFromPrimitive, IntoPrimitive, Eq, PartialEq)]
+#[repr(i8)]
+enum InitState {
+    Uninitd,
+    Initializing,
+    Initd,
+}
+
+/// Analogous to `std::sync::LazyLock`, but works on stable Rust, is no_std,
+/// only makes direct syscalls, etc.
+pub struct LazyLock<T, Init = fn() -> T> {
+    value: UnsafeCell<MaybeUninit<T>>,
+    init_state: AtomicI8,
+    initializer: UnsafeCell<Option<Init>>,
+}
+unsafe impl<T, Init> VirtualAddressSpaceIndependent for LazyLock<T, Init>
+where
+    T: VirtualAddressSpaceIndependent,
+    Init: VirtualAddressSpaceIndependent,
+{
+}
+// `T` must be `Sync`, since it will be shared across threads once initialized.
+// `Init` only needs to be `Send` since it'll only ever be accessed by the
+// single thread that performs the initialization.
+unsafe impl<T, Init> Sync for LazyLock<T, Init>
+where
+    T: Sync,
+    Init: Send,
+{
+}
+unsafe impl<T, Init> Send for LazyLock<T, Init>
+where
+    T: Send,
+    Init: Send,
+{
+}
+
+impl<T, Init> LazyLock<T, Init>
+where
+    Init: Producer<T>,
+{
+    // TODO: merge with `new` when loom's `UnsafeCell` supports const init.
+    #[cfg(not(loom))]
+    pub const fn const_new(initializer: Init) -> Self {
+        Self {
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            init_state: AtomicI8::new(InitState::Uninitd as i8),
+            initializer: UnsafeCell::new(Some(initializer)),
+        }
+    }
+
+    pub fn new(initializer: Init) -> Self {
+        Self {
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            init_state: AtomicI8::new(InitState::Uninitd.into()),
+            initializer: UnsafeCell::new(Some(initializer)),
+        }
+    }
+
+    fn init(&self, mut state: InitState) {
+        loop {
+            match state {
+                InitState::Uninitd => {
+                    match self.init_state.compare_exchange(
+                        InitState::Uninitd.into(),
+                        InitState::Initializing.into(),
+                        atomic::Ordering::Acquire,
+                        atomic::Ordering::Relaxed,
+                    ) {
```

Comment:
Nice catch; I'd forgotten to run loom after the latest changes.

Unfortunately the test still doesn't pass under loom. I think your suggestions are correct, but for completeness tried with and without - it fails either way. I'll try to debug some more tomorrow.

Reply:
Ah, I was accidentally using `core::atomic` instead of `crate::sync::atomic`, so the loom tests weren't recognizing that sync was happening at all. should have a fix in a bit...

Reply:
fixed. I ended up adding blocking via futex as well to get the loom tests to pass.

Reply:
Now that loom is working, if you switch the compare_exchange back to the original version does loom say anything? I'm curious to know whether it would catch that.

Reply:
Yup, it catches it :)